clippy_lints/src/struct_fields.rs

```diff
@@ -0,0 +1,168 @@
+    /// struct Cake {
+    ///     cake_sugar: u8,
+    ///     cake_flour: u8,
+    ///     cake_eggs: u8
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// struct Cake {
+    ///     sugar: u8,
+    ///     flour: u8,
+    ///     eggs: u8
+    /// }
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub STRUCT_FIELD_NAMES,
+    style,
+    "structs where all fields share a prefix/postfix"
+}
+
+pub struct StructFields {
+    threshold: u64,
+    avoid_breaking_exported_api: bool,
+}
+
+impl StructFields {
+    pub fn new(threshold: u64, avoid_breaking_exported_api: bool) -> StructFields {
+        StructFields {
+            threshold,
+            avoid_breaking_exported_api,
+        }
+    }
+}
+
+impl_lint_pass!(StructFields => [STRUCT_FIELD_NAMES]);
+
+fn is_snake_case(name: &str) -> bool {
+    name.chars().all(|c| !c.is_uppercase() || c == '_')
+}
+
+fn check_fields(cx: &LateContext<'_>, threshold: u64, fields: &[FieldDef<'_>], item_name: &str, item_span: Span) {
+    if (fields.len() as u64) < threshold {
+        return;
+    }
+
+    for field in fields {
+        if is_snake_case(field.ident.name.as_str()) {
+            check_struct_start(cx, item_name, field);
+            check_struct_end(cx, item_name, field);
+        }
+    }
+
+    let first_field = fields[0].ident.name.as_str();
+    let mut pre: Vec<&str> = first_field.split('_').collect();
+    let mut post = pre.clone();
+    post.reverse();
+    for field in fields {
+        let field_split: Vec<&str> = field.ident.name.as_str().split('_').collect();
+        if field_split.len() == 1 {
+            return;
+        }
+
+        pre = pre
+            .iter()
+            .zip(field_split.iter())
+            .take_while(|(a, b)| a == b)
+            .map(|e| *e.0)
+            .collect();
+        post = post
+            .iter()
+            .zip(field_split.iter().rev())
+            .take_while(|(a, b)| a == b)
+            .map(|e| *e.0)
+            .collect();
+    }
+    let prefix = pre.join("");
+    let postfix = post.join("");
+    if fields.len() > 1 {
+        let (what, value) = match (prefix.is_empty() || prefix == "_", postfix.is_empty()) {
+            (true, true) => return,
+            (false, _) => ("pre", prefix),
+            (true, false) => {
+                post.reverse();
+                ("post", postfix)
+            },
+        };
+        span_lint_and_help(
+            cx,
+            STRUCT_FIELD_NAMES,
+            item_span,
+            &format!("all fields have the same {what}fix: `{value}`"),
+            None,
+            &format!("remove the {what}fixes"),
+        );
+    }
+}
+
+fn check_struct_start(cx: &LateContext<'_>, item_name: &str, field: &FieldDef<'_>) {
+    let item_name_snake = to_snake_case(item_name);
+    let name = field.ident.name.as_str();
+    let item_name_chars = item_name_snake.chars().count();
+
+    if count_match_start(&item_name_snake, name).char_count == item_name_chars
```

Comment:
Is this different from just `name.starts_with(&item_name_snake)`? `starts_with` would be easier to read imo

Reply:
yes, it's much better with starts/ends_with. fixed in [654cea6](https://github.com/rust-lang/rust-clippy/pull/11496/commits/654cea62fdffba12178a41968457b69571df3cf3)